pkg/api/htpasswd.go

```diff
@@ -0,0 +1,203 @@
+// and caches all user passwords.
+func (hc *HtpasswdClient) Init() error {
+	credsFile, err := os.Open(hc.filepath)
+	if err != nil {
+		return fmt.Errorf("error occurred while opening creds-file: %w", err)
+	}
+	defer credsFile.Close()
+
+	hc.credMap.rw.Lock()
+	defer hc.credMap.rw.Unlock()
+
+	scanner := bufio.NewScanner(credsFile)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.Contains(line, ":") {
+			tokens := strings.Split(line, ":")
+			if len(tokens) == htpasswdValidTokensNumber {
+				hc.credMap.m[tokens[0]] = tokens[1]
+			}
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return fmt.Errorf("error occurred while reading creds-file: %w", err)
+	}
+
+	return nil
+}
+
+// Get returns the password associated with the login and a bool
+// indicating whether the login was found.
+// It does not check whether the user's password is correct.
+func (hc *HtpasswdClient) Get(login string) (string, bool) {
+	return hc.credMap.Get(login)
+}
+
+// Set sets the new password. It does not perform any checks,
+// the only error is possible is encryption error.
+func (hc *HtpasswdClient) Set(login, password string) error {
+	passphrase, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
+	if err != nil {
+		return fmt.Errorf("error occurred while cheking passwords: %w", err)
+	}
+
+	return hc.credMap.Set(login, string(passphrase))
+}
+
+// CheckPassword checks whether the user has a specified password.
+// It returns an error if the user is not found or passwords do not match,
+// and returns the nil on passwords match.
+func (hc *HtpasswdClient) CheckPassword(login, password string) error {
+	passwordHash, ok := hc.Get(login)
+	if !ok {
+		return zerr.ErrBadUser
+	}
+
+	err := bcrypt.CompareHashAndPassword([]byte(passwordHash), []byte(password))
+	if err != nil {
+		return zerr.ErrPasswordsDoNotMatch
+	}
+
+	return nil
+}
+
+// ChangePassword changes the user password.
+// It accepts user login, his supposed old password for verification and new password.
+func (hc *HtpasswdClient) ChangePassword(login, supposedOldPassword, newPassword string) error {
+	if len(newPassword) == 0 {
+		return zerr.ErrPasswordIsEmpty
+	}
+
+	oldPassphrase, ok := hc.credMap.Get(login)
+	if !ok {
+		return zerr.ErrBadUser
+	}
+
+	// given old password must match actual old password
+	if err := bcrypt.CompareHashAndPassword([]byte(oldPassphrase), []byte(supposedOldPassword)); err != nil {
+		return zerr.ErrOldPasswordIsWrong
+	}
+
+	// if passwords match, no need to update file and map, return nil as if operation is successful
+	if err := bcrypt.CompareHashAndPassword([]byte(oldPassphrase), []byte(newPassword)); err == nil {
+		return nil
+	}
+
+	// encrypt new password
+	newPassphrase, err := bcrypt.GenerateFromPassword([]byte(newPassword), bcrypt.DefaultCost)
+	if err != nil {
+		return fmt.Errorf("error occurred while encrypting new password: %w", err)
+	}
+
+	file, err := os.ReadFile(hc.filepath)
```

Comment:
We need a lock to block other users changing their passwords from reading/writing the file at the same time, and only one of the password changes ending up in the file.